mapie/multi_label_classification.py

```diff
@@ -0,0 +1,795 @@
+        y : NDArray of shape (n_samples, n_labels)
+            True labels.
+
+        Returns
+        -------
+        NDArray of shape (n_samples, n_lambdas)
+            Risks for each observation and each value of lambda.
+        """
+        y_pred_proba_repeat = np.repeat(
+            y_pred_proba,
+            self.n_lambdas,
+            axis=2
+        )
+
+        y_pred_th = (y_pred_proba_repeat > self.lambdas).astype(int)
+
+        y_repeat = np.repeat(y[..., np.newaxis], self.n_lambdas, axis=2)
+        risks = 1 - (
+            (y_pred_th * y_repeat).sum(axis=1) /
+            y.sum(axis=1)[:, np.newaxis]
+        )
+
+        return risks
+
+    def _get_r_hat_plus(
+        self,
+        bound: Optional[str],
+        delta: Optional[float],
+    ) -> Tuple[NDArray, NDArray]:
+        """Compute the upper bound of the loss for each lambda.
+
+        Parameters
+        ----------
+        bound : str
+            Bounds to compute. Either hoeffding, bernstein or wsr.
+        delta : float
+            Level of confidence.
+        sigma_init : float, optional
+            First variance in the sigma_hat array. The default
+            value is the same as in the paper implementation.
+
+            By default .25
+
+        Returns
+        -------
+        Tuple[NDArray, NDArray] of shape (n_lambdas, ) and (n_lambdas)
+            Average risk over all the obervations and upper bound of the risk.
+        """
+        r_hat = self.risks.mean(axis=0)
+        n_obs = len(self.risks)
+
+        if (self.method == "rcps") and (delta is not None):
+            if bound == "hoeffding":
+                r_hat_plus = (
+                    r_hat +
+                    np.sqrt((1 / (2 * n_obs)) * np.log(1 / delta))
+                )
+
+            elif bound == "bernstein":
+                sigma_hat_bern = np.var(r_hat, axis=0, ddof=1)
+                r_hat_plus = (
+                    r_hat +
+                    np.sqrt((sigma_hat_bern * 2 * np.log(2 / delta)) / n_obs) +
+                    (7 * np.log(2 / delta)) / (3 * (n_obs - 1))
+                )
+
+            else:
+                mu_hat = (
+                    (.5 + np.cumsum(self.risks, axis=0)) /
+                    (np.repeat(
+                        [range(1, n_obs + 1)],
+                        self.n_lambdas,
+                        axis=0
+                    ).T + 1)
+                )
+                sigma_hat = (
+                    (.25 + np.cumsum((self.risks - mu_hat)**2, axis=0)) /
+                    (np.repeat(
+                        [range(1, n_obs + 1)],
+                        self.n_lambdas,
+                        axis=0
+                    ).T + 1)
+                )
+                sigma_hat = np.concatenate(
+                    [
+                        np.full(
+                            (1, self.n_lambdas), fill_value=self.sigma_init
+                        ) * self.sigma_init, sigma_hat[:-1]
```

Comment:
self.sigma_init is an error I think! It was there because you use sigma_init X np.ones()...
Now you use np.full(fill_values=self.sigma_init)
you should delete the multiplication. If it is usefull, replace sigma_init by sigma_init**2
but I think this is an omission
Moreover, sigma_init is .25 or .25**2 ?

Reply:
sigman init is .25, I delete the omission

Reply:
sigman init is .25, I delete the omission